### PR TITLE
fix: [CDS-73520]: Adds support for validating github package registry artifact template

### PR DESCRIPTION
**Pull request by samarth-sharma31**
### Summary

This PR enables UI validation for Github Package registry artifact source template, Actually the validation for artifact templates of all types will be added in future but we are starting with github packages because it has internal nesting (on the basis of package type the fields rendered are different) and this cannot be handled via BE's schema validation. Whereas schema validation works for all the other artifact types, hence UI validation is not required for them at the moment

This handling is required for artifact template flow because we need to submit the form and get the errors when the template is saved in template studio.

#### Screenshots

https://github.com/harness/harness-core-ui/assets/96409598/d74f6bf3-014f-48d3-bed1-9f73137176e8

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Review by PratyushGar — APPROVED** [review]

**Review by rameshx — DISMISSED** [review]

File: src/modules/70-pipeline/components/ArtifactsSelection/hooks/useArtifactSelectionLastSteps.tsx [hunk 1]
```diff
@@ -106,7 +106,11 @@ export function useArtifactSelectionLastSteps(params: ArtifactSelectionLastSteps
       )
     case ENABLED_ARTIFACT_TYPES.GithubPackageRegistry:
       return (
-        <GithubPackageRegistry {...artifactLastStepProps} {...(shouldPassPrevStepData ? artifactPrevStepData : {})} />
+        <GithubPackageRegistryWithRef
```
Comment (chetanvekariya-harness) [inline src/modules/70-pipeline/components/ArtifactsSelection/hooks/useArtifactSelectionLastSteps.tsx:109]: Why this change is needed for only this artifact?

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx [hunk 2]
```diff
@@ -694,6 +697,7 @@ export function GithubPackageRegistry(
       isIdentifierAllowed
     ) as GithubPackageRegistryInitialValuesType
   }
+  const githubPackageRegistryFormRef = React.useRef<FormikProps<GithubPackageRegistryInitialValuesType> | null>()
```
Comment (chetanvekariya-harness) [inline src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx:700]: We already have yup validation and that should be able to take care of validation. Why do we need this special handling?

File: src/modules/70-pipeline/components/ArtifactsSelection/hooks/useArtifactSelectionLastSteps.tsx (hunk 1 above)
Comment (samarth-sharma31) [reply src/modules/70-pipeline/components/ArtifactsSelection/hooks/useArtifactSelectionLastSteps.tsx:109]: Actually the validation for artifact templates of all types will be added in future but we are starting with github packages because it has internal nesting (on the basis of package type the fields rendered are different) and this cannot be handled via BE's schema validation. Whereas schema validation works for all the other artifact types.

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx (hunk 2 above)
Comment (samarth-sharma31) [reply src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx:700]: This handling is required for artifact template flow because we need to submit the form and get the errors when the template is saved in template studio

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/__tests__/GithubPackageRegistry.test.tsx [hunk 3]
```diff
@@ -100,7 +167,56 @@ describe('GithubPackageRegistry tests', () => {
...
+        error: null,
+        loading: false
+      }
+    })
+    const { container } = render(
+      <TestWrapper>
+        <GithubPackageRegistryWithRef initialValues={commonInitialValues} {...props} context={ModalViewFor.SIDECAR} />
+      </TestWrapper>
+    )
+    expect(container).toMatchSnapshot()
```
Comment (rameshx) [inline src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/__tests__/GithubPackageRegistry.test.tsx:201]: Could this snapshot be replaced with a few assertions?

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx [hunk 4]
```diff
@@ -175,6 +177,7 @@ function FormComponent({
     }))
   }, [versionDetails?.data])
 
+  /* istanbul ignore next */
```
Comment (chetanvekariya-harness) [inline src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx:180]: nit: remove this

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx [hunk 5]
```diff
@@ -249,18 +252,20 @@ function FormComponent({
...
-                    packageType: value.value,
-                    packageName: formik.values?.spec?.packageName === RUNTIME_INPUT_VALUE ? RUNTIME_INPUT_VALUE : '',
-                    version: formik.values?.spec?.version === RUNTIME_INPUT_VALUE ? RUNTIME_INPUT_VALUE : ''
-                  }
-                })
-              }}
+              onChange={
+                /* istanbul ignore next */ value => {
+                  formik.setValues({
+                    ...omit(formik.values, ['packageSource']),
```
Comment (chetanvekariya-harness) [inline src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx:258]: nit: no need of this

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx [hunk 6]
```diff
@@ -810,6 +817,18 @@ export function GithubPackageRegistry(
     )
   })
 
+  /* istanbul ignore next */ React.useImperativeHandle(formikRef, () => ({
```
Comment (chetanvekariya-harness) [inline src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx:820]: We can remove this.

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx [hunk 7]
```diff
@@ -832,9 +851,12 @@ export function GithubPackageRegistry(
         }}
       >
         {formik => {
+          githubPackageRegistryFormRef.current = formik
```
Comment (chetanvekariya-harness) [inline src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx:854]: formikRef = formik can be done here.

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/__tests__/GithubPackageRegistry.test.tsx [hunk 8]
```diff
@@ -91,16 +84,139 @@ describe('GithubPackageRegistry tests', () => {
...
+        loading: false
+      }
+    })
     const { container } = render(
       <TestWrapper>
-        <GithubPackageRegistry initialValues={commonInitialValues} {...props} />
+        <GithubPackageRegistryWithRef initialValues={commonInitialValues} {...props} />
+      </TestWrapper>
+    )
+    expect(container).toMatchSnapshot()
```
Comment (chetanvekariya-harness) [inline src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/__tests__/GithubPackageRegistry.test.tsx:167]: Let's avoid snapshot testing.

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/__tests__/GithubPackageRegistry.test.tsx (hunk 3 above)
Comment (chetanvekariya-harness) [reply src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/__tests__/GithubPackageRegistry.test.tsx:201]: +1

**Review by chetanvekariya-harness — APPROVED** [review]

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/__tests__/GithubPackageRegistry.test.tsx [hunk 9]
```diff
@@ -243,6 +359,118 @@ describe('GithubPackageRegistry tests', () => {
...
+    await waitFor(() => {
+      expect(props.handleSubmit).toBeCalled()
+      expect(props.handleSubmit).toHaveBeenCalledWith({
+        spec: {
+          connectorRef: 'testConnector',
+          packageType: 'maven',
+          org: 'sample org',
+          packageName: 'testPackage',
+          versionRegex: '*.zip',
+          groupId: '<+input>',
```
Comment (chetanvekariya-harness) [inline src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/__tests__/GithubPackageRegistry.test.tsx:466]: nit: RUNTIME_INPUT_VALUE constant can be used.

**Review by chetanvekariya-harness — APPROVED** [review]

**Review by rameshx — APPROVED** [review]

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx (hunk 7 above)
Comment (samarth-sharma31) [reply src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx]: done

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx (hunk 6 above)
Comment (samarth-sharma31) [reply src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx]: done

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx (hunk 5 above)
Comment (samarth-sharma31) [reply src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx]: done

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx (hunk 4 above)
Comment (samarth-sharma31) [reply src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/GithubPackageRegistry.tsx]: removed

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/__tests__/GithubPackageRegistry.test.tsx [hunk 10]
```diff
@@ -91,16 +117,105 @@ describe('GithubPackageRegistry tests', () => {
...
+        error: null,
+        loading: false
+      }
+    })
+    const { container } = render(
+      <TestWrapper>
+        <GithubPackageRegistryWithRef initialValues={commonInitialValues} {...props} context={ModalViewFor.SIDECAR} />
+      </TestWrapper>
+    )
+    const queryByNameAttribute = (name: string): HTMLElement | null => queryByAttribute('name', container, name)
```
[the remark this answers is not in the document]
Comment (samarth-sharma31) [reply src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/__tests__/GithubPackageRegistry.test.tsx]: yes got it, done, replaced usages

File: src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/__tests__/GithubPackageRegistry.test.tsx (hunk 9 above)
Comment (samarth-sharma31) [reply src/modules/70-pipeline/components/ArtifactsSelection/ArtifactRepository/ArtifactLastSteps/GithubPackageRegistry/__tests__/GithubPackageRegistry.test.tsx]: yes agreed, done

**Review by rameshx — APPROVED** [review]

**Review by PratyushGar — APPROVED** [review]

**Review by chetanvekariya-harness — APPROVED** [review]
